// wechaty-puppet-mock/dist/cjs/src/mock/generator.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateSentence = exports.generateRoomPayload = exports.generateMessagePayloadTo = exports.generateMessagePayloadRoom = exports.generateImageFileBox = exports.generateContactPayload = void 0;
const faker_1 = __importDefault(require("faker"));
const cuid_1 = __importDefault(require("cuid"));
const wechaty_puppet_1 = require("wechaty-puppet");
const generateContactPayload = () => ({
    address: faker_1.default.address.streetAddress(),
    alias: undefined,
    avatar: faker_1.default.image.avatar(),
    city: faker_1.default.address.city(),
    friend: true,
    gender: wechaty_puppet_1.ContactGender.Male,
    id: (0, cuid_1.default)(),
    name: faker_1.default.name.findName(),
    phone: [
        faker_1.default.phone.phoneNumber(),
    ],
    province: faker_1.default.address.state(),
    signature: faker_1.default.lorem.sentence(),
    star: false,
    type: wechaty_puppet_1.ContactType.Individual,
    weixin: undefined,
});
exports.generateContactPayload = generateContactPayload;
const generateImageFileBox = () => wechaty_puppet_1.FileBox.fromUrl(faker_1.default.image.avatar());
exports.generateImageFileBox = generateImageFileBox;
const generateRoomPayload = (...contactIdList) => {
    const maxNum = Math.max(500, contactIdList.length);
    const roomNum = Math.floor(maxNum * Math.random());
    const shuffledList = contactIdList.sort(() => Math.random() - 0.5);
    const memberIdList = shuffledList.slice(0, roomNum);
    const payload = {
        adminIdList: [],
        avatar: faker_1.default.image.avatar(),
        id: (0, cuid_1.default)() + '@chatroom',
        memberIdList,
        ownerId: undefined,
        topic: faker_1.default.lorem.word(),
    };
    return payload;
};
exports.generateRoomPayload = generateRoomPayload;
const generateMessagePayloadTo = () => ({
    fromId: (0, cuid_1.default)(),
    id: (0, cuid_1.default)(),
    text: faker_1.default.lorem.sentence(),
    timestamp: Date.now(),
    toId: (0, cuid_1.default)(),
    type: wechaty_puppet_1.MessageType.Text,
});
exports.generateMessagePayloadTo = generateMessagePayloadTo;
const generateMessagePayloadRoom = () => ({
    fromId: (0, cuid_1.default)(),
    id: (0, cuid_1.default)(),
    mentionIdList: [],
    roomId: (0, cuid_1.default)() + '@chatroom',
    text: faker_1.default.lorem.sentence(),
    timestamp: Date.now(),
    type: wechaty_puppet_1.MessageType.Text,
});
exports.generateMessagePayloadRoom = generateMessagePayloadRoom;
const generateSentence = () => faker_1.default.lorem.sentence();
exports.generateSentence = generateSentence;
