// wechaty/dist/cjs/src/mod.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Wechaty = exports.VERSION = exports.UrlLink = exports.Tag = exports.ScanStatus = exports.RoomInvitation = exports.Room = exports.qrcodeValueToImageUrl = exports.Money = exports.Moment = exports.MiniProgram = exports.Message = exports.MemoryCard = exports.log = exports.IoClient = exports.Location = exports.Image = exports.Friendship = exports.FileBox = exports.Favorite = exports.ContactSelf = exports.Contact = exports.config = void 0;
/**
 *   Wechaty Chatbot SDK - https://github.com/wechaty/wechaty
 *
 *   @copyright 2016 Huan LI (李卓桓) <https://github.com/huan>, and
 *                   Wechaty Contributors <https://github.com/wechaty>.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */
const wechaty_puppet_1 = require("wechaty-puppet");
Object.defineProperty(exports, "ScanStatus", { enumerable: true, get: function () { return wechaty_puppet_1.ScanStatus; } });
Object.defineProperty(exports, "FileBox", { enumerable: true, get: function () { return wechaty_puppet_1.FileBox; } });
Object.defineProperty(exports, "MemoryCard", { enumerable: true, get: function () { return wechaty_puppet_1.MemoryCard; } });
Object.defineProperty(exports, "log", { enumerable: true, get: function () { return wechaty_puppet_1.log; } });
const config_js_1 = require("./config.js");
Object.defineProperty(exports, "config", { enumerable: true, get: function () { return config_js_1.config; } });
Object.defineProperty(exports, "qrcodeValueToImageUrl", { enumerable: true, get: function () { return config_js_1.qrcodeValueToImageUrl; } });
Object.defineProperty(exports, "VERSION", { enumerable: true, get: function () { return config_js_1.VERSION; } });
/**
 * We need to put `Wechaty` at the beginning of this file for import
 * because we have circular dependencies between `Puppet` & `Wechaty`
 */
const wechaty_js_1 = require("./wechaty.js");
Object.defineProperty(exports, "Wechaty", { enumerable: true, get: function () { return wechaty_js_1.Wechaty; } });
const mod_js_1 = require("./user/mod.js");
Object.defineProperty(exports, "Contact", { enumerable: true, get: function () { return mod_js_1.Contact; } });
Object.defineProperty(exports, "ContactSelf", { enumerable: true, get: function () { return mod_js_1.ContactSelf; } });
Object.defineProperty(exports, "Favorite", { enumerable: true, get: function () { return mod_js_1.Favorite; } });
Object.defineProperty(exports, "Friendship", { enumerable: true, get: function () { return mod_js_1.Friendship; } });
Object.defineProperty(exports, "Image", { enumerable: true, get: function () { return mod_js_1.Image; } });
Object.defineProperty(exports, "Message", { enumerable: true, get: function () { return mod_js_1.Message; } });
Object.defineProperty(exports, "MiniProgram", { enumerable: true, get: function () { return mod_js_1.MiniProgram; } });
Object.defineProperty(exports, "Moment", { enumerable: true, get: function () { return mod_js_1.Moment; } });
Object.defineProperty(exports, "Money", { enumerable: true, get: function () { return mod_js_1.Money; } });
Object.defineProperty(exports, "Room", { enumerable: true, get: function () { return mod_js_1.Room; } });
Object.defineProperty(exports, "RoomInvitation", { enumerable: true, get: function () { return mod_js_1.RoomInvitation; } });
Object.defineProperty(exports, "Tag", { enumerable: true, get: function () { return mod_js_1.Tag; } });
Object.defineProperty(exports, "UrlLink", { enumerable: true, get: function () { return mod_js_1.UrlLink; } });
Object.defineProperty(exports, "Location", { enumerable: true, get: function () { return mod_js_1.Location; } });
const io_client_js_1 = require("./io-client.js");
Object.defineProperty(exports, "IoClient", { enumerable: true, get: function () { return io_client_js_1.IoClient; } });
