// wechaty/dist/cjs/src/user/favorite.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.wechatifyFavorite = exports.Favorite = void 0;
/**
 *   Wechaty Chatbot SDK - https://github.com/wechaty/wechaty
 *
 *   @copyright 2016 Huan LI (李卓桓) <https://github.com/huan>, and
 *                   Wechaty Contributors <https://github.com/wechaty>.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */
const wechaty_puppet_1 = require("wechaty-puppet");
const guard_wechatify_class_js_1 = require("./guard-wechatify-class.js");
class Favorite {
    static get wechaty() { return (0, guard_wechatify_class_js_1.throwWechatifyError)(this); }
    get wechaty() { return (0, guard_wechatify_class_js_1.throwWechatifyError)(this.constructor); }
    static list() {
        return [];
    }
    /**
     * Get tags for all favorites
     *
     * @static
     * @returns {Promise<Tag[]>}
     * @example
     * const tags = await wechaty.Favorite.tags()
     */
    static async tags() {
        wechaty_puppet_1.log.verbose('Favorite', 'static tags() for %s', this);
        // TODO:
        // try {
        //   const tagIdList = await this.puppet.tagFavoriteList()
        //   const tagList = tagIdList.map(id => this.wechaty.Tag.load(id))
        //   return tagList
        // } catch (e) {
        //   log.error('Favorite', 'static tags() exception: %s', e.message)
        //   return []
        // }
        return [];
    }
    /*
     * @hideconstructor
     */
    constructor() {
        guard_wechatify_class_js_1.guardWechatifyClass.call(this, Favorite);
    }
    async tags() {
        // TODO: implmente this method
        return [];
    }
    async findAll() {
        //
    }
}
exports.Favorite = Favorite;
function wechatifyFavorite(wechaty) {
    wechaty_puppet_1.log.verbose('Favorite', 'wechatifyFavorite(%s)', wechaty);
    class WechatifiedFavorite extends Favorite {
        static get wechaty() { return wechaty; }
        get wechaty() { return wechaty; }
    }
    return WechatifiedFavorite;
}
exports.wechatifyFavorite = wechatifyFavorite;
