// wechaty-puppet/dist/esm/src/mod.js
import { FileBoxType, FileBox, } from 'file-box';
import { MemoryCard, } from 'memory-card';
import { StateSwitch, } from 'state-switch';
import { VERSION, log, } from './config.js';
import { looseInstanceOfFileBox, looseInstanceOfPuppet, } from './loose-instance-of.js';
import { Puppet, } from './puppet.js';
import { ContactGender, ContactType, } from './schemas/contact.js';
import { ScanStatus, } from './schemas/event.js';
import { FriendshipType, FriendshipSceneType, } from './schemas/friendship.js';
import { ImageType, } from './schemas/image.js';
import { MessageType, } from './schemas/message.js';
import { throwUnsupportedError, } from './throw-unsupported-error.js';
import { PayloadType } from './schemas/payload.js';
import { CHAT_EVENT_DICT, PUPPET_EVENT_DICT, YOU, } from './schemas/puppet.js';
export { CHAT_EVENT_DICT, ContactGender, ContactType, FileBox, FileBoxType, FriendshipSceneType, FriendshipType, ImageType, log, looseInstanceOfFileBox, looseInstanceOfPuppet, MemoryCard, MessageType, PayloadType, PUPPET_EVENT_DICT, Puppet, ScanStatus, StateSwitch, throwUnsupportedError, VERSION, YOU, };
