// wechaty-grpc/dist/cjs/src/mod.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VERSION = exports.StringValue = exports.puppet = exports.proto = exports.openApi = exports.grpc = void 0;
const wrappers_pb_js_1 = __importDefault(require("google-protobuf/google/protobuf/wrappers_pb.js"));
const grpc = __importStar(require("@grpc/grpc-js"));
exports.grpc = grpc;
const proto = __importStar(require("./proto.js"));
exports.proto = proto;
const openApi = __importStar(require("./openapi.js"));
exports.openApi = openApi;
const cjs_js_1 = require("./cjs.js");
Object.defineProperty(exports, "puppet", { enumerable: true, get: function () { return cjs_js_1.puppet; } });
const config_js_1 = require("./config.js");
Object.defineProperty(exports, "VERSION", { enumerable: true, get: function () { return config_js_1.VERSION; } });
const { StringValue, } = wrappers_pb_js_1.default;
exports.StringValue = StringValue;
