// wechaty/dist/esm/src/user/url-link.js
/**
 *   Wechaty Chatbot SDK - https://github.com/wechaty/wechaty
 *
 *   @copyright 2016 Huan LI (李卓桓) <https://github.com/huan>, and
 *                   Wechaty Contributors <https://github.com/wechaty>.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */
import Url from 'url';
import { log, } from 'wechaty-puppet';
import { openGraph, } from '../helper-functions/impure/open-graph.js';
import { guardWechatifyClass, throwWechatifyError, } from './guard-wechatify-class.js';
class UrlLink {
    /*
     * @hideconstructor
     */
    constructor(payload) {
        this.payload = payload;
        log.verbose('UrlLink', 'constructor()');
        guardWechatifyClass.call(this, UrlLink);
    }
    static get wechaty() { return throwWechatifyError(this); }
    get wechaty() { return throwWechatifyError(this.constructor); }
    /**
     *
     * Create from URL
     *
     */
    static async create(url) {
        log.verbose('UrlLink', 'create(%s)', url);
        const meta = await openGraph(url);
        let description;
        let imageUrl;
        let title;
        if (meta.image) {
            if (typeof meta.image === 'string') {
                imageUrl = meta.image;
            }
            else if (Array.isArray(meta.image)) {
                imageUrl = meta.image[0];
            }
            else {
                if (Array.isArray(meta.image.url)) {
                    imageUrl = meta.image.url[0];
                }
                else if (meta.image.url) {
                    imageUrl = meta.image.url;
                }
            }
        }
        if (Array.isArray(meta.title)) {
            title = meta.title[0];
        }
        else {
            title = meta.title;
        }
        if (Array.isArray(meta.description)) {
            description = meta.description[0];
        }
        else if (meta.description) {
            description = meta.description;
        }
        else {
            description = title;
        }
        if (!imageUrl || !description) {
            throw new Error(`imageUrl(${imageUrl}) or description(${description}) not found!`);
        }
        if (!imageUrl.startsWith('http')) {
            const resolvedUrl = new Url.URL(imageUrl, url);
            imageUrl = resolvedUrl.toString();
        }
        const payload = {
            description,
            thumbnailUrl: imageUrl,
            title,
            url,
        };
        return new this(payload);
    }
    toString() {
        return `UrlLink<${this.payload.url}>`;
    }
    url() {
        return this.payload.url;
    }
    title() {
        return this.payload.title;
    }
    thumbnailUrl() {
        return this.payload.thumbnailUrl;
    }
    description() {
        return this.payload.description;
    }
}
function wechatifyUrlLink(wechaty) {
    log.verbose('UrlLink', 'wechatifyUrlLink(%s)', wechaty);
    class WechatifiedUrlLink extends UrlLink {
        static get wechaty() { return wechaty; }
        get wechaty() { return wechaty; }
    }
    return WechatifiedUrlLink;
}
export { UrlLink, wechatifyUrlLink, };
